File: September-LeetCoding-Challenge/12-Combination-Sum-III/Combination-Sum-III.cs
public class Solution {
        public IList<IList<int>> CombinationSum3(int k, int n)
        {
            var res = new List<IList<int>>();
            var tmp = new List<int>();
            Backtrack(res, tmp, k, n, 1);
            return res;
        }

        public void Backtrack(List<IList<int>> res, List<int> tmp, int k, int n, int idx)
        {
            if (k == tmp.Count && n == 0)
            {
                res.Add(new List<int>(tmp)); //Deep clone
                return;
            }

            for (int i = idx; i <= n && k > 0 && n > 0; i++)
            {
                if(i>9) break;
                tmp.Add(i);
                Backtrack(res, tmp, k, n - i, i + 1);
                tmp.RemoveAt(tmp.Count - 1); //Remove last item when it doesn't find the correct combo
            }
        }
}

File: September-LeetCoding-Challenge/12-Combination-Sum-III/Combination-Sum-III.rb
def combination_sum3(k, n)
    path = []
    res = []
    bt(k, n, 1, path, res)
    res
    
end

def bt(k, n, s, path, res)
    if k == 0
        if n == 0
            res << path.clone
            return
        end
    end
    
    s.upto(9) do |i|
        return if i > n
        path << i
        bt(k-1, n-i, i+1, path, res)
        path.pop
    end
end